### PR TITLE
fix(amp-carousel): Guard against empty slides carousel

### PR DESCRIPTION
**Pull request by erwinmombay**
Closes https://github.com/ampproject/amphtml/issues/2213

### CODE REVIEW

**Comment by erwinmombay** [comment]
@dvoytenko PTAL. let me know if you want to assert to publisher dev about empty carousels as well.

File: extensions/amp-carousel/0.1/slides.js [hunk 1]
```diff
@@ -271,7 +290,7 @@ export class AmpSlides extends BaseCarousel {
    * @private
    */
   onSwipeStart_(unusedSwipe) {
```
Comment (jridgewell) [inline extensions/amp-carousel/0.1/slides.js]: How about `this.currentIndex_ < this.slides.length - 1`?

File: extensions/amp-carousel/0.1/slides.js [hunk 2]
```diff
@@ -392,7 +411,9 @@ export class AmpSlides extends BaseCarousel {
     if (this.isLooping_) {
       return true;
     }
-    return this.currentIndex_ != this.slides_.length - 1;
+    return this.currentIndex_ != this.slides_.length - 1 &&
```
Comment (jridgewell) [inline extensions/amp-carousel/0.1/slides.js]: How about `this.currentIndex_ < this.slides.length - 1`?

File: extensions/amp-carousel/0.1/slides.js [hunk 3]
```diff
@@ -45,6 +45,10 @@ export class AmpSlides extends BaseCarousel {
 
     /** @private {!Array<!Element>} */
     this.slides_ = this.getRealChildren();
+
+    AMP.assert(this.slides_.length < 1,
```
Comment (jridgewell) [inline extensions/amp-carousel/0.1/slides.js]: Did you mean `>=`?

File: extensions/amp-carousel/0.1/slides.js [hunk 4]
```diff
@@ -63,21 +67,29 @@ export class AmpSlides extends BaseCarousel {
...
     return Promise.resolve();
   }
 
   /** @override */
   viewportCallback(inViewport) {
-    this.updateInViewport(this.slides_[this.currentIndex_], inViewport);
-    this.tryAutoplay_(1, true);
+    const curSlide = this.curSlide_();
+    if (curSlide) {
+      this.updateInViewport(this.slides_[this.currentIndex_], inViewport);
```
Comment (jridgewell) [inline extensions/amp-carousel/0.1/slides.js]: `this.curSlide_()`

File: extensions/amp-carousel/0.1/slides.js (hunk 3 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: haha yes, premature push.

File: extensions/amp-carousel/0.1/slides.js (hunk 4 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: done.

File: extensions/amp-carousel/0.1/slides.js (hunk 2 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: done.

**Comment by erwinmombay** [comment]
@dvoytenko PTAL

File: extensions/amp-carousel/0.1/slides.js [hunk 5]
```diff
@@ -63,21 +67,29 @@ export class AmpSlides extends BaseCarousel {
...
 
   /** @override */
   goCallback(dir, animate) {
     const newIndex = this.nextIndex_(dir);
-    if (newIndex != this.currentIndex_) {
+    // Guard again NaN by checking if greater than or equal to zero
+    // since we can't have negative indexes anyways.
+    if (newIndex >= 0 && newIndex != this.currentIndex_) {
       const newSlide = this.slides_[newIndex];
       const oldSlide = this.slides_[this.currentIndex_];
```
Comment (jridgewell) [inline extensions/amp-carousel/0.1/slides.js]: `this.curSlide_()`

Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: done.

**Comment by jridgewell** [comment]
LGTM

**Comment by erwinmombay** [comment]
merging this build was green [here](https://travis-ci.org/ampproject/amphtml/builds/112098430), but restarted for somereason

**Pull request merged**